In [291]:
#importing libaries

In [364]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import os
#os.listdir()

# Loading datasets

In [365]:
with open("spambase.names")as filex:
    reading =filex.readlines()

In [366]:
words= [word.split(':')[0]for word in flines[33:]]

In [367]:
words

['word_freq_make',
 'word_freq_address',
 'word_freq_all',
 'word_freq_3d',
 'word_freq_our',
 'word_freq_over',
 'word_freq_remove',
 'word_freq_internet',
 'word_freq_order',
 'word_freq_mail',
 'word_freq_receive',
 'word_freq_will',
 'word_freq_people',
 'word_freq_report',
 'word_freq_addresses',
 'word_freq_free',
 'word_freq_business',
 'word_freq_email',
 'word_freq_you',
 'word_freq_credit',
 'word_freq_your',
 'word_freq_font',
 'word_freq_000',
 'word_freq_money',
 'word_freq_hp',
 'word_freq_hpl',
 'word_freq_george',
 'word_freq_650',
 'word_freq_lab',
 'word_freq_labs',
 'word_freq_telnet',
 'word_freq_857',
 'word_freq_data',
 'word_freq_415',
 'word_freq_85',
 'word_freq_technology',
 'word_freq_1999',
 'word_freq_parts',
 'word_freq_pm',
 'word_freq_direct',
 'word_freq_cs',
 'word_freq_meeting',
 'word_freq_original',
 'word_freq_project',
 'word_freq_re',
 'word_freq_edu',
 'word_freq_table',
 'word_freq_conference',
 'char_freq_;',
 'char_freq_(',
 'char_freq_[',
 '

In [368]:
df = pd.read_csv('spambase (1).data',names= words)

df.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0.00            0.64               0.64            0.0          0.32   
0.21            0.28               0.50            0.0          0.14   
0.06            0.00               0.71            0.0          1.23   
0.00            0.00               0.00            0.0          0.63   
0.00            0.00               0.00            0.0          0.63   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0.00           0.00            0.00              0.00                0.00   
0.21           0.28            0.21              0.07                0.00   
0.06           0.19            0.19              0.12                0.64   
0.00           0.00            0.31              0.63                0.31   
0.00           0.00            0.31              0.63                0.31   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0.00             0.00            0.00  ...                  0.00        0.000   
0.21             0.94            0.21  ...                  0.00        0.132   
0.06             0.25            0.38  ...                  0.01        0.143   
0.00             0.63            0.31  ...                  0.00        0.137   
0.00             0.63            0.31  ...                  0.00        0.135   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0.00          0.0        0.778        0.000        0.000        3.756   
0.21          0.0        0.372        0.180        0.048        5.114   
0.06          0.0        0.276        0.184        0.010        9.821   
0.00          0.0        0.137        0.000        0.000        3.537   
0.00          0.0        0.135        0.000        0.000        3.537   

      capital_run_length_average  capital_run_length_longest  \
0.00                          61                         278   
0.21                         101                        1028   
0.06                         485                        2259   
0.00                          40                         191   
0.00                          40                         191   

      capital_run_length_total  
0.00                         1  
0.21                         1  
0.06                         1  
0.00                         1  
0.00                         1  

[5 rows x 57 columns]

# FEATURE EXPLORATIONS

In [369]:
df.describe()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
count     4601.000000        4601.000000    4601.000000   4601.000000   
mean         0.213015           0.280656       0.065425      0.312223   
std          1.290575           0.504143       1.395151      0.672513   
min          0.000000           0.000000       0.000000      0.000000   
25%          0.000000           0.000000       0.000000      0.000000   
50%          0.000000           0.000000       0.000000      0.000000   
75%          0.000000           0.420000       0.000000      0.380000   
max         14.280000           5.100000      42.810000     10.000000   

       word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
count    4601.000000     4601.000000       4601.000000         4601.000000   
mean        0.095901        0.114208          0.105295            0.090067   
std         0.273824        0.391441          0.401071            0.278616   
min         0.000000        0.000000          0.000000            0.000000   
25%         0.000000        0.000000          0.000000            0.000000   
50%         0.000000        0.000000          0.000000            0.000000   
75%         0.000000        0.000000          0.000000            0.000000   
max         5.880000        7.270000         11.110000            5.260000   

       word_freq_order  word_freq_mail  ...  word_freq_conference  \
count      4601.000000     4601.000000  ...           4601.000000   
mean          0.239413        0.059824  ...              0.038575   
std           0.644755        0.201545  ...              0.243471   
min           0.000000        0.000000  ...              0.000000   
25%           0.000000        0.000000  ...              0.000000   
50%           0.000000        0.000000  ...              0.000000   
75%           0.160000        0.000000  ...              0.000000   
max          18.180000        2.610000  ...              4.385000   

       char_freq_;  char_freq_(  char_freq_[  char_freq_!  char_freq_$  \
count  4601.000000  4601.000000  4601.000000  4601.000000  4601.000000   
mean      0.139030     0.016976     0.269071     0.075811     0.044238   
std       0.270355     0.109394     0.815672     0.245882     0.429342   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.065000     0.000000     0.000000     0.000000     0.000000   
75%       0.188000     0.000000     0.315000     0.052000     0.000000   
max       9.752000     4.081000    32.478000     6.003000    19.829000   

       char_freq_#  capital_run_length_average  capital_run_length_longest  \
count  4601.000000                 4601.000000                 4601.000000   
mean      5.191515                   52.172789                  283.289285   
std      31.729449                  194.891310                  606.347851   
min       1.000000                    1.000000                    1.000000   
25%       1.588000                    6.000000                   35.000000   
50%       2.276000                   15.000000                   95.000000   
75%       3.706000                   43.000000                  266.000000   
max    1102.500000                 9989.000000                15841.000000   

       capital_run_length_total  
count               4601.000000  
mean                   0.394045  
std                    0.488698  
min                    0.000000  
25%                    0.000000  
50%                    0.000000  
75%                    1.000000  
max                    1.000000  

[8 rows x 57 columns]

In [370]:
#CHECKING  FOR NULL VALUES IN OUR DATASET
df.isnull().sum()

word_freq_make                0
word_freq_address             0
word_freq_all                 0
word_freq_3d                  0
word_freq_our                 0
word_freq_over                0
word_freq_remove              0
word_freq_internet            0
word_freq_order               0
word_freq_mail                0
word_freq_receive             0
word_freq_will                0
word_freq_people              0
word_freq_report              0
word_freq_addresses           0
word_freq_free                0
word_freq_business            0
word_freq_email               0
word_freq_you                 0
word_freq_credit              0
word_freq_your                0
word_freq_font                0
word_freq_000                 0
word_freq_money               0
word_freq_hp                  0
word_freq_hpl                 0
word_freq_george              0
word_freq_650                 0
word_freq_lab                 0
word_freq_labs                0
word_freq_telnet              0
word_fre

In [371]:
words

['word_freq_make',
 'word_freq_address',
 'word_freq_all',
 'word_freq_3d',
 'word_freq_our',
 'word_freq_over',
 'word_freq_remove',
 'word_freq_internet',
 'word_freq_order',
 'word_freq_mail',
 'word_freq_receive',
 'word_freq_will',
 'word_freq_people',
 'word_freq_report',
 'word_freq_addresses',
 'word_freq_free',
 'word_freq_business',
 'word_freq_email',
 'word_freq_you',
 'word_freq_credit',
 'word_freq_your',
 'word_freq_font',
 'word_freq_000',
 'word_freq_money',
 'word_freq_hp',
 'word_freq_hpl',
 'word_freq_george',
 'word_freq_650',
 'word_freq_lab',
 'word_freq_labs',
 'word_freq_telnet',
 'word_freq_857',
 'word_freq_data',
 'word_freq_415',
 'word_freq_85',
 'word_freq_technology',
 'word_freq_1999',
 'word_freq_parts',
 'word_freq_pm',
 'word_freq_direct',
 'word_freq_cs',
 'word_freq_meeting',
 'word_freq_original',
 'word_freq_project',
 'word_freq_re',
 'word_freq_edu',
 'word_freq_table',
 'word_freq_conference',
 'char_freq_;',
 'char_freq_(',
 'char_freq_[',
 '

In [372]:
df.dtypes

word_freq_make                float64
word_freq_address             float64
word_freq_all                 float64
word_freq_3d                  float64
word_freq_our                 float64
word_freq_over                float64
word_freq_remove              float64
word_freq_internet            float64
word_freq_order               float64
word_freq_mail                float64
word_freq_receive             float64
word_freq_will                float64
word_freq_people              float64
word_freq_report              float64
word_freq_addresses           float64
word_freq_free                float64
word_freq_business            float64
word_freq_email               float64
word_freq_you                 float64
word_freq_credit              float64
word_freq_your                float64
word_freq_font                float64
word_freq_000                 float64
word_freq_money               float64
word_freq_hp                  float64
word_freq_hpl                 float64
word_freq_ge

<B># Extracting Independent and dependent VariableA</B>



In [373]:
x=df.iloc[0:,0:56].values
y=df.iloc[:,-1].values

In [374]:
x

array([[6.400e-01, 6.400e-01, 0.000e+00, ..., 3.756e+00, 6.100e+01,
        2.780e+02],
       [2.800e-01, 5.000e-01, 0.000e+00, ..., 5.114e+00, 1.010e+02,
        1.028e+03],
       [0.000e+00, 7.100e-01, 0.000e+00, ..., 9.821e+00, 4.850e+02,
        2.259e+03],
       ...,
       [0.000e+00, 3.000e-01, 0.000e+00, ..., 1.404e+00, 6.000e+00,
        1.180e+02],
       [0.000e+00, 0.000e+00, 0.000e+00, ..., 1.147e+00, 5.000e+00,
        7.800e+01],
       [0.000e+00, 6.500e-01, 0.000e+00, ..., 1.250e+00, 5.000e+00,
        4.000e+01]])

In [375]:
#Splitting the dataset into training and test set.
#from here we can see that after spliting, we gave our test data =25%, meaning that our remaing data which is training has 75%
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state= 144)


In [376]:
x_train

array([[  0.   ,   0.   ,   0.   , ...,   1.36 ,   3.   ,  34.   ],
       [  0.   ,   0.   ,   0.   , ...,   4.333,  13.   ,  78.   ],
       [  0.25 ,   0.   ,   0.   , ...,   3.181,  32.   , 210.   ],
       ...,
       [  0.   ,   0.   ,   0.   , ...,   3.333,  10.   ,  30.   ],
       [  0.   ,   0.4  ,   0.   , ...,   1.727,  11.   , 190.   ],
       [  0.   ,   0.   ,   0.   , ...,   2.446,  11.   , 115.   ]])

* as we can see from this training data is not scaled since we want to compare  the difference between scaled and unscaled data

In [377]:
# checking for rows and columns
x_train.shape

(3680, 56)

In [378]:
y_train.shape

(3680,)

<b> Fitting K-nearest neighbor to our training data by choosing kvalue=8 </b>



In [379]:
from sklearn.neighbors import KNeighborsClassifier

In [434]:
knn_model=KNeighborsClassifier(n_neighbors=5)

In [435]:
# fitting the model
knn_model.fit(x_train,y_train)

KNeighborsClassifier()

In [436]:
## predicting the Test set Results
y_pred=knn_model.predict(x_test)


# EVALUATION

*  I am going to test Test Accuracy of the result for KNN Algorithm by  using both  the Confusion matrix and accuracy score, and classification report 

In [437]:
#
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test,y_pred)

array([[482,  76],
       [108, 255]], dtype=int64)

 * out of 100: TP+TN= 482+255=737 FP+FN=482+255=737

* we can interpret that − 108+76= 184 (Correct Output), − and 1108+76= 184(Incorrect Output).

In [439]:
#Accuracy = (TP+TN)/total
from sklearn.metrics import accuracy_score
print("accuracy score:", accuracy_score(y_test,y_pred))

accuracy score: 0.8002171552660152


• Accuracy measures the proportion of true results to total cases.
− Aim for a high accuracy rate.
• Accuracy is a common evaluation metric for classification problems.
It’s the number of correct predictions made as a ratio of all predictions
made.

In [426]:
from sklearn.metrics import classification_report
matrix = classification_report(y_test,y_pred)
print("classifcation report:",matrix)


classifcation report:               precision    recall  f1-score   support

           0       0.81      0.85      0.83       558
           1       0.75      0.69      0.72       363

    accuracy                           0.79       921
   macro avg       0.78      0.77      0.77       921
weighted avg       0.79      0.79      0.79       921



In [427]:
#fitting model with decesion tree algorithm and try compare its accuracy with the previous model algorithm which is KNN
# defining model

In [428]:
from sklearn.tree import DecisionTreeClassifier
Tree_model=DecisionTreeClassifier()
Tree_model.fit(x_train,y_train)

DecisionTreeClassifier()

In [429]:
#predicting
y_predict_tree=Tree_model.predict(x_test)

In [431]:
accuracy_score(y_test,y_predict_tree)

0.9044516829533116

In [432]:
from sklearn.metrics import classification_report
matrix = classification_report(y_test,y_predict_tree)
print("classifcation report:",matrix)


classifcation report:               precision    recall  f1-score   support

           0       0.92      0.92      0.92       558
           1       0.88      0.88      0.88       363

    accuracy                           0.90       921
   macro avg       0.90      0.90      0.90       921
weighted avg       0.90      0.90      0.90       921



In [433]:
from sklearn.metrics import confusion_matrix
accuracy =confusion_matrix(y_test,y_predict_tree)
accuracy

array([[514,  44],
       [ 44, 319]], dtype=int64)

# After working with unscaled data,I am going to work with scaled data and compare their accuracy from KNN and decision tree algorithms



In [249]:
#data scaling 
from sklearn.preprocessing import MinMaxScaler
minimax_model=MinMaxScaler()
x_train_scale=minimax_model.fit_transform(x_train)
x_train_scale

array([[0.        , 0.        , 0.        , ..., 0.00032683, 0.00090785,
        0.00208333],
       [0.        , 0.        , 0.        , ..., 0.00302587, 0.00544712,
        0.00486111],
       [0.017507  , 0.        , 0.        , ..., 0.00198003, 0.01407172,
        0.01319444],
       ...,
       [0.        , 0.        , 0.        , ..., 0.00211802, 0.00408534,
        0.00183081],
       [0.        , 0.07843137, 0.        , ..., 0.00066001, 0.00453926,
        0.01193182],
       [0.        , 0.        , 0.        , ..., 0.00131276, 0.00453926,
        0.00719697]])

In [233]:
x_test_scale=minimax_model.fit_transform(x_test)
x_test_scale

array([[0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.00039757],
       [0.        , 0.        , 0.        , ..., 0.00748   , 0.00140168,
        0.02415267],
       [0.        , 0.        , 0.        , ..., 0.003732  , 0.00060072,
        0.00566544],
       ...,
       [0.        , 0.        , 0.        , ..., 0.005468  , 0.00230276,
        0.0646059 ],
       [0.        , 0.        , 0.        , ..., 0.018152  , 0.00400481,
        0.00705695],
       [0.02871148, 0.04405286, 0.        , ..., 0.0026    , 0.00140168,
        0.0172945 ]])

 # KNN algorithm in scaled  data

In [440]:
# fitting the model using KNN
knn_model.fit(x_train_scale,y_train)

KNeighborsClassifier()

In [441]:

#predicting model
y_pred_scale=knn_model.predict(x_test_scale)


In [442]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test,y_pred_scale)

array([[516,  42],
       [ 60, 303]], dtype=int64)

In [443]:
accuracy_score(y_test,y_pred_scale)

0.8892508143322475

   # Decision tree in scaled  data

In [444]:
from sklearn.tree import DecisionTreeClassifier
Tree_model=DecisionTreeClassifier()
Tree_model.fit(x_train_scale,y_train)

DecisionTreeClassifier()

In [445]:
#predicting
y_predict_tree=Tree_model.predict(x_test_scale)

In [446]:
# checking accuracy for scaled data

from sklearn.metrics import confusion_matrix
confusion_matrix(y_test,y_pred_scale)

array([[516,  42],
       [ 60, 303]], dtype=int64)

In [447]:
accuracy_score(y_test,y_predict_tree)

0.7915309446254072

# Performing  feauture importance 


In [404]:
# feature selection
from sklearn.feature_selection import SelectKBest, chi2

xnew = SelectKBest(chi2, k=23).fit_transform(x,y)

# split into test and training
xtrain,xtest,ytrain,ytest= train_test_split(xnew,y,test_size=0.25,random_state=42)

# scaling the data
from sklearn.preprocessing import StandardScaler
scale = StandardScaler()
xtrain = scale.fit_transform(xtrain)
xtest = scale.transform(xtest)

# fitting the model
knn=KNeighborsClassifier(n_neighbors=13)
knn.fit(xtrain, ytrain)
ypred = knn.predict(xtest)
ypredtrain = knn.predict(xtrain)

# accuracy
accuracy = accuracy_score(ytest,ypred)
accuracytrain = accuracy_score(ytrain,ypredtrain)
accuracytrain, accuracy

(0.9243478260869565, 0.9131190269331017)